### DISCUSSION TITLE
Transition actions and weird TypeScript errors

### DISCUSSION THREAD

**Discussion by Perlkonig** [statelyai/xstate#1383] [General]
Yes, it's me again. Thank you for your assistance thus far. Hopefully you can help me here too.

As per discussion #1377, I've been refactoring my machine to switch from enter/exit actions (where appropriate) to transition actions. As per discusstion #1359, what I've been doing is using @xstate/immer and `assign` to create mutators that are defined outside of the machine and that get referenced by variable name in the machine definition, instead of by string.

Everything works fine when I do this as an entry/exit action, but moving it to the transition is causing TypeScript errors that are making my head explode. Below is the minimal example that mirrors as closely as possible how my current code is structured, including all the extra TypeScript typings recommended in the docs.

```ts
import { Machine, interpret } from "xstate";
import { assign } from '@xstate/immer';
import { immerable } from "immer";

class MyContextClass {
    [immerable] = true;
    readonly seedStart: string;
    // scratch?: string;
    constructor (seed: string) {
        this.seedStart = seed;
    }
}

interface StateSchema {
    states: {
        waiting: {};
        elsewhere: {};
    }
}

type LeaveEvent = { type: "LEAVE" }
type CampEvent = { type: "CAMP", days: number };
type Events = LeaveEvent | CampEvent;

const tickTracker = assign<MyContextClass, Events>((context, event) => {
    // tslint:disable-next-line: no-console
    console.log(event);
});

const leaveArea = assign<MyContextClass, Events>((context, event) => {
    // tslint:disable-next-line: no-console
    console.log(event);
});

const camp = assign<MyContextClass, Events>((context, event) => {
    // tslint:disable-next-line: no-console
    console.log(event);
});

const machine = Machine<MyContextClass, StateSchema, Events>(
    {
        initial: "waiting",
        states: {
            waiting: {
                entry: [tickTracker],
                on: {
                    LEAVE: {
                        target: "elsewhere",
                        actions: [leaveArea]
                    },
                    CAMP: {
                        actions: [camp]
                    }
                }
            },
            elsewhere: {
            }
        }
    }
);
const s = interpret(machine.withContext(new MyContextClass("seed")));
s.start();
s.send("CAMP", {days: 2});
s.send("LEAVE");
/* Expected output:
    { type: 'xstate.init' }
    { type: 'CAMP', days: 2 }
    { type: 'LEAVE' }
*/
```

If you run that as is, everything works great. But see that commented-out line `scratch?: string`? Uncomment it and then try running it. You will get a TS2322 error with a long string of assignment errors. An excerpt below:

```txt
Type '{ target: string; actions: AssignAction<MyContextClass, Events>[]; }' is not assignable to type 'SingleOrArray<string | StateNode<MyContextClass, any, LeaveEvent, any> | TransitionConfig<MyContextClass, LeaveEvent>>'.
  Types of property 'actions' are incompatible.
    Type 'AssignAction<MyContextClass, Events>[]' is not assignable to type 'SingleOrArray<Action<MyContextClass, LeaveEvent>>'.
      Type 'AssignAction<MyContextClass, Events>[]' is not assignable to type 'Action<MyContextClass, LeaveEvent>[]'.
        Type 'AssignAction<MyContextClass, Events>' is not assignable to type 'Action<MyContextClass, LeaveEvent>'.
          Type 'AssignAction<MyContextClass, Events>' is not assignable to type 'AssignAction<Required<MyContextClass>, LeaveEvent>'.
            Types of property 'exec' are incompatible.
              Type 'ActionFunction<MyContextClass, Events>' is not assignable to type 'ActionFunction<Required<MyContextClass>, LeaveEvent>'.
                Types of parameters 'meta' and 'meta' are incompatible.
                  Type 'ActionMeta<Required<MyContextClass>, LeaveEvent>' is not assignable to type 'ActionMeta<MyContextClass, Events>'.
                    The types of 'state.configuration' are incompatible between these types.
                      Type 'StateNode<Required<MyContextClass>, any, LeaveEvent, any>[]' is not assignable to type 'StateNode<MyContextClass, any, Events, any>[]'.
                        Type 'StateNode<Required<MyContextClass>, any, LeaveEvent, any>' is not assignable to type 'StateNode<MyContextClass, any, Events, any>'.
                          The types of 'config.on' are incompatible between these types.
                            Type 'TransitionsConfig<Required<MyContextClass>, LeaveEvent>' is not assignable to type 'TransitionsConfig<MyContextClass, Events>'.
                              Type 'TransitionsConfigMap<Required<MyContextClass>, LeaveEvent>' is not assignable to type 'TransitionsConfig<MyContextClass, Events>'.
                                Type 'TransitionsConfigMap<Required<MyContextClass>, LeaveEvent>' is not assignable to type 'TransitionsConfigMap<MyContextClass, Events>'.
                                  Type 'TransitionsConfigMap<Required<MyContextClass>, LeaveEvent>' is not assignable to type '{ ''?: SingleOrArray<string | TransitionConfig<MyContextClass, Events> | StateNode<MyContextClass, any, Events, any>>; }'.
                                    Types of property '''' are incompatible.
                                      Type 'SingleOrArray<string | StateNode<Required<MyContextClass>, any, LeaveEvent, any> | TransitionConfig<Required<MyContextClass>, LeaveEvent>>' is not assignable to type 'SingleOrArray<string | TransitionConfig<MyContextClass, Events> | StateNode<MyContextClass, any, Events, any>>'.
                                        Type 'StateNode<Required<MyContextClass>, any, LeaveEvent, any>' is not assignable to type 'SingleOrArray<string | TransitionConfig<MyContextClass, Events> | StateNode<MyContextClass, any, Events, any>>'.
```

So you'll see that the entry/exit node works fine, but moving it to an transition action blows everything up. And I cannot for the life of me, after over 90 minutes of reading and isolating the problem, figure out why or how to fix it.

Any insights? Is this something I should submit as an issue? I really appreciate your time and patience.

**Comment by Perlkonig** [comment]
I have a CodeSandbox link here (https://codesandbox.io/s/xstate-transition-action-issue-b3hru?file=/index.ts) but I cannot for the life of me figure out how to get to an actual command line. But if I export it to ZIP and extract on my dev machine, I can type `npx ts-node index.ts` to run the experiment (after `npm i`, of course).

**Reply by Perlkonig** [reply]
I would be grateful if someone could try replicating the issue. If it's just me, then I'd sure like to know. Thanks again.

**Reply by davidkpiano** [reply]
Can you try with the latest XState 4.12.0 (just released)?

**Reply by Perlkonig** [reply]
That produces errors even with the line commented out:

```txt
⨯ Unable to compile TypeScript:
index.ts:45:7 - error TS2322: Type '{ LEAVE: { target: string; actions: AssignAction<MyContextClass, Events>[]; }; CAMP: { actions: AssignAction<MyContextClass, Events>[]; }; }' is not assignable to type 'TransitionsConfigMap<MyContextClass, Events> | TransitionsConfigArray<MyContextClass, Events> | undefined'.
  Types of property 'LEAVE' are incompatible.
    Type '{ target: string; actions: AssignAction<MyContextClass, Events>[]; }' is not assignable to type 'SingleOrArray<string | StateNode<MyContextClass, any, LeaveEvent, { value: any; context: MyContextClass; }> | TransitionConfig<MyContextClass, LeaveEvent> | undefined>'.
      Types of property 'actions' are incompatible.
        Type 'AssignAction<MyContextClass, Events>[]' is not assignable to type 'string | ActionObject<MyContextClass, LeaveEvent> | ActionFunction<MyContextClass, LeaveEvent> | Action<...>[] | undefined'.
          Type 'AssignAction<MyContextClass, Events>[]' is not assignable to type 'Action<MyContextClass, LeaveEvent>[]'.
            Type 'AssignAction<MyContextClass, Events>' is not assignable to type 'Action<MyContextClass, LeaveEvent>'.
              Type 'AssignAction<MyContextClass, Events>' is not assignable to type 'string'.

45       on: {
         ~~

  node_modules/xstate/lib/types.d.ts:294:5
    294     on?: TransitionsConfig<TContext, TEvent>;
            ~~
    The expected type comes from property 'on' which is declared here on type 'StateNodeConfig<MyContextClass, {}, Events>'
index.ts:58:21 - error TS2345: Argument of type 'StateNode<MyContextClass, any, StateSchema, { value: any; context: MyContextClass; }>' is not assignable to parameter of type 'StateMachine<MyContextClass, any, EventObject, { value: any; context: MyContextClass; }>'.
  The types of 'config.invoke' are incompatible between these types.
    Type 'StateMachine<any, any, any, { value: any; context: any; }> | InvokeConfig<MyContextClass, StateSchema> | (StateMachine<any, any, any, { ...; }> | InvokeConfig<...>)[] | undefined' is not assignable to type 'StateMachine<any, any, any, { value: any; context: any; }> | InvokeConfig<MyContextClass, EventObject> | (StateMachine<any, any, any, { ...; }> | InvokeConfig<...>)[] | undefined'.
      Type 'InvokeConfig<MyContextClass, StateSchema>' is not assignable to type 'StateMachine<any, any, any, { value: any; context: any; }> | InvokeConfig<MyContextClass, EventObject> | (StateMachine<any, any, any, { ...; }> | InvokeConfig<...>)[] | undefined'.
        Type 'InvokeConfig<MyContextClass, StateSchema>' is not assignable to type 'InvokeConfig<MyContextClass, EventObject>'.
          Property 'states' is missing in type 'EventObject' but required in type 'StateSchema'.

58 const s = interpret(machine.withContext(new MyContextClass("seed")));
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  index.ts:15:3
    15   states: {
         ~~~~~~
    'states' is declared here.
```

**Reply by davidkpiano** [reply]
Do you have a working CodeSandbox link? The one you posted doesn't show the error. Or even a repo.

**Reply by Perlkonig** [reply]
What do you mean "working"? I'm new to the whole codesandbox thing, but [the link above, duplicated here,](https://codesandbox.io/s/xstate-transition-action-issue-b3hru) contains the four files. What I've been doing is exporting that as a ZIP to my dev machine.

- With v4.11.0 and the line commented out, everything works.
- With v4.11.0 and the line included, I get the original errors.
- With 4.12.0 and the line commented out, I get the new errors I just posted.

**Reply by davidkpiano** [reply]
"Working" as in able to accurately and consistently reproduce the issue without any other superfluous errors. The `tsconfig.json` file is throwing an error in the above CodeSandbox, and when fixing that, the CodeSandbox shows no errors.

Also ensure that your local TypeScript version is up to date.

**Reply by Perlkonig** [reply]
While it is true that CodeSandbox doesn't show any issues, the code still does not compile (and my local VSCode install shows the errors I have described). I run `npx ts-node index.ts` to see the output from the program. I haven't found a way to execute a command line within CodeSandbox other than the browser console.

I have fixed the `tsconfig.json` in the sandbox. My `tsc` is indeed v3.9.7.

**Reply by Andarist** [reply]
I've just tried this and:
- I don't see the problem on the CodeSandbox
- I don't see it in the VSCode either, after downloading that sandbox and trying out locally
- `npx ts-node index.ts` doesn't report any problems for me either

I would love to investigate this further, but don't know what I can do without having a way to experience this on my own :s

**Reply by Perlkonig** [reply]
So uncommenting that line `// scratch?: string;` doesn't cause any errors to be thrown? OK, then. Thanks for that. Now to figure out what on my local machine could cause this.

**Reply by Andarist** [reply]
Sorry - I was going really quickly through this discussion and I haven't read your post carefully. I have not uncommented this line - when I do it then the error is reported, both on the sandbox and locally.

For the future - it would be better if a repro case would demonstrate the issue right away, without extra steps though 😉 We handle multiple issues, discussions, etc simultaneously and it's hard to give each one maximum attention. If a repro case is provided I usually just jump into it straight away, not always reading the associated comments.

Gonna look into why this might happen tomorrow.

**Reply by Perlkonig** [reply]
I concur. I guess I just wanted to demonstrate that I had isolated it to a specific line of code. I'll be more clear in the future. Thanks for your time. Do you want me to post this as an issue?

**Reply by davidkpiano** [reply]
> Do you want me to post this as an issue?

Yes please. This is a tricky issue because we're stretching the limits of TypeScript inference here.

If you update your CodeSandbox to XState 4.12, you'll notice the main problem is that TS is complaining that `Action<SomeContext, SpecificEvent>` is not assignable to `Action<SomeContext, AnyEvent>`, and that `assign(...)` action defined is the latter. Narrowing it to `assign<SomeContext, SpecificEvent>(...)` is a "workaround" but you shouldn't have to do that, ideally.

**Reply by Perlkonig** [reply]
Filed. Thanks again for your time.
